matchtree.go

```diff
@@ -983,14 +991,33 @@ func (d *indexData) newMatchTree(q query.Q, opt matchTreeOpt) (matchTree, error)
 		optCopy := opt
 		optCopy.DisableWordMatchOptimization = true
 
-		subMT, err := d.newMatchTree(s.Expr, optCopy)
+		anchored := false
+		expr := s.Expr
+		switch e := s.Expr.(type) {
+		case *query.Regexp:
+			pattern := e.Regexp.String()
+			if strings.HasPrefix(pattern, "^") && strings.HasSuffix(pattern, "$") {
+				pattern = pattern[1 : len(pattern)-1]
+				parsedPattern, err := syntax.Parse(pattern, e.Regexp.Flags)
+				if err != nil {
+					return nil, err
+				}
+				eCopy := *e
+				eCopy.Regexp = parsedPattern
+				expr = &eCopy
+				anchored = true
+			}
+		}
+
+		subMT, err := d.newMatchTree(expr, optCopy)
```

Comment:
I looked at putting this logic into `newMatchTree`, but the problem is that anchors aren't meaningful until we get to the symbol layer because we want the anchors to match the start and end of the symbol name which are not available lower in the tree.